kb_sdk_plus.md

```diff
@@ -1,10 +1,98 @@
 
+## Why?
+
+The KBase SDK is a core piece of infrastructure, but it has become exceedingly difficult to
+maintain and expand as the test suite is exceptionally complex and currently does not pass.
+As such, large changes to the codebase are risky and no updates have been released for 5+ years.
+Furthermore, due to outdated dependencies, the SDK will not run on modern machines.
+
+To rectify this while still supporting older apps, we have elected to fork the
+https://github.com/kbase/kb_sdk repo to `kb_sdk_plus`. Version 1.2.1 is the last `kb_sdk`
+release and will be kept as is. Future SDK development will occur in `kb_sdk_plus`.
+
+## Why not kb_sdk2?
+
+`<repo>2` generally signifies a greenfield rewrite of a codebase in KBase,
+which `kb_sdk_plus` is not. It's forking a codebase to leave the original codebase untouched
+and runnable while we make changes to the forked codebase.
+
+## Guidelines
+
+* Backwards incompatible changes are expected in some respects:
+  * R code generation will be removed
+    * Unused in KBase or anywhere else as far as we know other than one test app
+  * Perl code generation will almost certainly be removed
+    * Perl apps can continue to be supported with the original SDK, or wrapped with
+      Python / Java
+      * https://github.com/boriel/perlfunc for instance
+* Otherwise `kb_sdk_plus` needs to be backwards compatible with `kb_sdk`
+  * It must work with the current KBase backend services
+  * Python and Java apps must continue to be supported
+    * It might be possible to make backwards incompatible changes that are fixed
+      with a recompile of an app or a simple migration guide.
+      This would need careful thought and implementation.
+* Generally simplify the codebase where possible and document where needed
+      
+## Prior to releasing for general development
+
+* Get the tests to pass locally
+  * Remove R support, probably remove Perl support
+  * Do we still need Javascript support?
+* Remove the git submodules and `submodules_hacks` directory
+* See if the `sdkbase` directory is still needed, remove if not
+* Try to remove the `lib` directory
+* Remove `JAR_DEPS`, `JAR_DEPS_BIN`, and `DEPENDENCIES`
+* Understand the purpose of the code in the `entrypoint` script and simplify if possible
+* Use the dockerized JobRunner callback server for tests
+  * E.g. remove the Java callback server implementation
+* Ensure tests are running against EE2 vs. EE1 (e.g. the `KBaseJobService.spec`
+  file is suspicious)
+* Fix docker version issues (see Dakota's patch: https://github.com/kbase/kb_sdk_patch)
+* Convert the build to Gradle
+  * Remove `build.xml`, `travis.yml` and `.classpath`
+* Update to Java 11 at least. Test with newer Java versions if possible
+* Get the tests running in GHA with coverage, dependabot, and trivy
+  * At least look into running macOS and windows tests in GHA
+    * If not, clearly document manual testing requirements before a release
+    * Do we need to support windows?
```

Comment:
That might be a question for Ben.

Reply:
I was going to ask exec later down the line